File: src/main/java/com/raychenon/leetcode/StringCompression.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 1/3/2020
 * https://leetcode.com/problems/string-compression/
 */
public class StringCompression {

    /**
     * time complexity : O(n)
     * space complexity: O(1) , use the chars input to store the data
     *
     * @param chars
     * @return
     */
    public int compress(char[] chars) {
        int anchor = 0, write = 0;
        int size = chars.length;
        for (int i = 0; i < size; i++) {
            if (i + 1 == size || chars[i + 1] != chars[i]) {
                chars[write++] = chars[anchor];
                if (i > anchor) {
                    for (char c : ("" + (i - anchor + 1)).toCharArray()) {
                        chars[write++] = c;
                    }
                }
                anchor = i + 1;
            }
        }
        return write;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/StringCompression.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 26/2/2020
 * https://leetcode.com/problems/string-compression/
 */
object StringCompression {

    fun compress(chars: CharArray): Int {

        if (chars.isEmpty() || chars.size == 0) {
            return 0
        }

        var str = StringBuilder()
        var left = 0
        var right = 0
        val n = chars.size

        while (left < n) {
            while ((right < n) && chars.get(left) == chars.get(right)) {
                right++
            }
            str.append(chars.get(left))
            val freq = right - left
            if (freq > 1) {
                str.append(freq)
            }
            left = right
        }

        // print("compress, str = ${str.toString()} \n")
        return str.length
    }
}